plugin/plugins/filesystem/nodes/utils.go

```diff
@@ -0,0 +1,278 @@
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+func doClone(names []string) bool {
+	l := len(names)
+	if l < 1 {
+		return true
+	}
+	//TODO: double check the spec to make sure dot handling is correct
+	// we may only want to clone on ".." if we're a root
+	if pc := names[0]; l == 1 && (pc == ".." || pc == "." || pc == "") {
+		return true
+	}
+	return false
+}
+
+//TODO: rename this and/or extend
+// it only does some of the stat and not what people probably expect
+func coreStat(ctx context.Context, dirEnt *p9.Dirent, core coreiface.CoreAPI, path corepath.Path) error {
+	if ipldNode, err := core.ResolveNode(ctx, path); err != nil {
+		return err
+	} else {
+		return ipldStat(dirEnt, ipldNode)
+	}
+}
+
+func coreGetAttr(ctx context.Context, attr *p9.Attr, attrMask p9.AttrMask, core coreiface.CoreAPI, path corepath.Path) (err error) {
+	ipldNode, err := core.ResolveNode(ctx, path)
+	if err != nil {
+		return err
+	}
+	ufsNode, err := unixfs.ExtractFSNode(ipldNode)
+	if err != nil {
+		return err
+	}
+
+	if attrMask.Mode {
+	attr.Mode = IRXA //TODO: this should probably be the callers responsability; just document that permissions should be set afterwards or something
+	attr.Mode |= unixfsTypeTo9Mode(ufsNode.Type())
+	}
+
+	if attrMask.Blocks{
+	if bs := ufsNode.BlockSizes(); len(bs) != 0 {
+		attr.BlockSize = bs[0] //NOTE: this value is to be used as a hint only; subsequent child block size may differ
+	}
+
+	//TODO [eventualy]: switch off here for handling of time metadata in new format standard
+	timeStamp(attr, attrMask)
+}
+
+if attrMask.Size {
+	attr.Size, attrMask.Size = ufsNode.FileSize(), true
+}
+
+if attrMask.RDev {
+	switch path.Namespace() {
+	case "ipfs":
+		attr.RDev, attrMask.RDev = dIPFS, true
+		//case "ipns":
+		//attr.RDev, attrMask.RDev = dIPNS, true
+		//etc.
+	}
+}
+
+	return nil
+}
+
+func ipldStat(dirEnt *p9.Dirent, node ipld.Node) error {
+	ufsNode, err := unixfs.ExtractFSNode(node)
+
+	if err != nil {
+		return err
+	}
+
+	nodeType := unixfsTypeTo9Mode(ufsNode.Type()).QIDType()
+
+	dirEnt.Type = nodeType
+	dirEnt.QID.Type = nodeType
+	dirEnt.QID.Path = cidToQPath(node.Cid())
+
+	return nil
+}
+
+func cidToQPath(cid cid.Cid) uint64 {
+	hasher := fnv.New64a()
+	if _, err := hasher.Write(cid.Bytes()); err != nil {
+		panic(err)
+	}
+	return hasher.Sum64()
+}
+
+func coreLs(ctx context.Context, corePath corepath.Path, core coreiface.CoreAPI) (<-chan coreiface.DirEntry, error) {
+
+	//FIXME: asyncContext hangs on reset
+	//asyncContext := deriveTimerContext(ctx, 10*time.Second)
+	asyncContext := ctx
+
+	coreChan, err := core.Unixfs().Ls(asyncContext, corePath)
+	if err != nil {
+		//asyncContext.Cancel()
+		return nil, err
+	}
+
+	oStat, err := core.Object().Stat(asyncContext, corePath)
+	if err != nil {
+		return nil, err
+	}
+
+	relayChan := make(chan coreiface.DirEntry)
+	go func() {
+		//defer asyncContext.Cancel()
+		defer close(relayChan)
+
+		for i := 0; i != oStat.NumLinks; i++ {
```

Comment:
This won't work with sharded directories.

Reply:
Can't we just wait for `Ls` to close `coreChan`?

Reply:
I changed how IPFS Readdir logic works, so that the system checks and stores the coreChan on Open(), and progresses on it during Readdir(), in a forward-only fashion.
https://github.com/ipfs/go-ipfs/blob/059c218a76149c270f6cd11d21a4a9dbd9172bae/plugin/plugins/filesystem/nodes/ipfs.go#L103